iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/DiskGuardian.java

```diff
@@ -0,0 +1,246 @@
+    }
+
+    @Override
+    public String toString() {
+      return String.format(
+          "[Raft Log Summary]: group=%s, total size=%d, files=%s", gid, totalSize, logFiles);
+    }
+  }
+
+  /** use a {@link MemoizedSupplier} here to avoid this reference leak */
+  private final MemoizedSupplier<RatisConsensus> serverRef;
+
+  /** whether a specific RaftGroup need to take snapshot during next check period */
+  private final Map<RaftGroupId, AtomicBoolean> snapshotFlag = new ConcurrentHashMap<>();
+
+  /** Raft Log disk usage summary statistics */
+  private final Map<RaftGroupId, RaftLogSummary> bookkeeper = new ConcurrentHashMap<>();
+
+  /** Registered checkers */
+  private final Map<TimeDuration, List<Predicate<RaftLogSummary>>> snapshotArbitrators =
+      new HashMap<>();
+
+  private final ScheduledExecutorService workerThread =
+      IoTDBThreadPoolFactory.newSingleThreadScheduledExecutor(
+          ThreadName.RATIS_BG_DISK_GUARDIAN.getName());
+  private final AtomicBoolean isStopped = new AtomicBoolean(false);
+  private final long daemonIntervalMs;
+
+  DiskGuardian(Supplier<RatisConsensus> server, RatisConfig config) {
+    this.serverRef = MemoizedSupplier.valueOf(server);
+    this.daemonIntervalMs = config.getImpl().getCheckAndTakeSnapshotInterval();
+  }
+
+  void start() {
+    // first schedule the snapshot daemon
+    ScheduledExecutorUtil.safelyScheduleWithFixedDelay(
+        workerThread, this::snapshotDaemon, 0L, daemonIntervalMs, TimeUnit.SECONDS);
+
+    // then schedule all checker daemons
+    snapshotArbitrators.forEach(
+        (interval, checkers) ->
+            ScheduledExecutorUtil.safelyScheduleWithFixedDelay(
+                workerThread,
+                () -> checkerDaemon(checkers),
+                0L,
+                interval.toLong(TimeUnit.SECONDS),
+                TimeUnit.SECONDS));
+  }
+
+  void stop() throws InterruptedException {
+    if (isStopped.compareAndSet(true, false)) {
+      workerThread.shutdown();
+      workerThread.awaitTermination(5, TimeUnit.SECONDS);
+    }
+  }
+
+  /** call this method to register checkers before {@link #start()} */
+  void registerChecker(Predicate<RaftLogSummary> checker, TimeDuration interval) {
+    final List<Predicate<RaftLogSummary>> checkers =
+        snapshotArbitrators.computeIfAbsent(interval, i -> new CopyOnWriteArrayList<>());
+    checkers.add(checker);
+  }
+
+  /** periodically woken up. Will take a snapshot if a flag is set. */
+  private void snapshotDaemon() {
+    if (isStopped.get()) {
+      return;
+    }
+    for (RaftGroupId groupId : serverRef.get().getServer().getGroupIds()) {
+      if (getSnapshotFlag(groupId).get()) {
+        try {
+          serverRef.get().triggerSnapshot(Utils.fromRaftGroupIdToConsensusGroupId(groupId));
+          final boolean flagCleared = snapshotFlag.get(groupId).compareAndSet(true, false);
+          Preconditions.assertTrue(flagCleared);
```

Comment:
Although this shouldn't be false in theory, maybe it's nicer to log an error in case something goes wrong than to throw an exception?

Reply:
It's your call, captain.